bundles/org.openhab.core.thing/src/test/java/org/openhab/core/thing/binding/BaseDynamicStateDescriptionProviderTest.java

```diff
@@ -0,0 +1,112 @@
+import org.openhab.core.thing.events.ChannelDescriptionChangedEvent;
+import org.openhab.core.thing.events.ChannelDescriptionChangedEvent.CommonChannelDescriptionField;
+import org.openhab.core.thing.link.ItemChannelLinkRegistry;
+import org.openhab.core.types.StateOption;
+import org.osgi.framework.BundleContext;
+
+/**
+ * Tests for {@link BaseDynamicStateDescriptionProvider}.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.WARN)
+class BaseDynamicStateDescriptionProviderTest {
+
+    private static final ThingTypeUID THING_TYPE_UID = new ThingTypeUID("binding:type");
+    private static final ThingUID THING_UID = new ThingUID(THING_TYPE_UID, "id");
+    private static final ChannelUID CHANNEL_UID = new ChannelUID(THING_UID, "channel");
+
+    @Mock
+    EventPublisher mockEventPublisher;
+
+    @Mock
+    ItemChannelLinkRegistry mockItemChannelLinkRegistry;
+
+    class TestBaseDynamicStateDescriptionProvider extends BaseDynamicStateDescriptionProvider {
+
+        public TestBaseDynamicStateDescriptionProvider() {
+            this.bundleContext = mock(BundleContext.class);
+            this.eventPublisher = mockEventPublisher;
+            this.itemChannelLinkRegistry = mockItemChannelLinkRegistry;
+        }
+    };
+
+    private BaseDynamicStateDescriptionProvider subject;
+
+    @BeforeEach
+    public void setup() {
+        when(mockItemChannelLinkRegistry.getLinkedItemNames(CHANNEL_UID)).thenReturn(Set.of("item1", "item2"));
+
+        subject = new TestBaseDynamicStateDescriptionProvider();
+    }
+
+    @Test
+    public void setStatePatternPublishesEvent() {
+        subject.setStatePattern(CHANNEL_UID, "%s");
+
+        ArgumentCaptor<Event> capture = ArgumentCaptor.forClass(Event.class);
+        verify(mockEventPublisher, times(1)).post(capture.capture());
+
+        Event event = capture.getValue();
+        assertTrue(event instanceof ChannelDescriptionChangedEvent);
+        ChannelDescriptionChangedEvent cdce = (ChannelDescriptionChangedEvent) event;
+        assertEquals(CommonChannelDescriptionField.PATTERN, cdce.getField());
+
+        // check the event is not published again
+        subject.setStatePattern(CHANNEL_UID, "%s");
+
+        verify(mockEventPublisher, times(1)).post(capture.capture());
+    }
+
+    @Test
+    public void setStateOptionsPublishesEvent() {
+        subject.setStateOptions(CHANNEL_UID, List.of(new StateOption("offline", "Offline")));
+
+        ArgumentCaptor<Event> capture = ArgumentCaptor.forClass(Event.class);
+        verify(mockEventPublisher, times(1)).post(capture.capture());
+
+        Event event = capture.getValue();
+        assertTrue(event instanceof ChannelDescriptionChangedEvent);
+        ChannelDescriptionChangedEvent cdce = (ChannelDescriptionChangedEvent) event;
+        assertEquals(CommonChannelDescriptionField.STATE_OPTIONS, cdce.getField());
+
```

Comment:
Suggestion: you could additionally checked the linked items and the list of options in the event.